### PR TITLE
Validation fu added to examples/structural_mechanics/crash/train.py

### PR DESCRIPTION
**Pull request by dakhare-creator**
# PhysicsNeMo Pull Request

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: examples/structural_mechanics/crash/conf/training/default.yaml [hunk 1]
```diff
@@ -27,9 +27,12 @@ max_workers_preprocessing: 64  # Maximum parallel workers
 
 num_time_steps: 14
 num_training_samples: 8
+num_validation_samples: 8
 start_lr: 0.0001
 end_lr: 0.0000003
 epochs: 10000
+validate_every_n_epochs: 10
```
Comment (mnabian) [inline examples/structural_mechanics/crash/conf/training/default.yaml]: Maybe change this to validation_freq?

File: examples/structural_mechanics/crash/conf/training/default.yaml [hunk 2]
```diff
@@ -27,9 +27,12 @@ max_workers_preprocessing: 64  # Maximum parallel workers
...
+save_ckpt_every_n_epochs: 10
```
Comment (mnabian) [inline examples/structural_mechanics/crash/conf/training/default.yaml]: And change this to save_chckpoint_freq?

File: examples/structural_mechanics/crash/train.py [hunk 3]
```diff
@@ -109,6 +110,58 @@ def __init__(self, cfg: DictConfig, logger0: RankZeroLoggingWrapper):
...
+                cfg.training.num_validation_samples
+                // self.num_validation_replicas
+                * self.num_validation_replicas
+            )
+            logger0.info(f"Number of validation samples: {self.num_validation_samples}")
+
+            # Create a validation dataset
+            val_cfg = self.cfg.datapipe
+            with open_dict(val_cfg):  # or open_dict(cfg) to open the whole tree
+                val_cfg.data_dir = self.cfg.inference.raw_data_dir_test
```
Comment (mnabian) [inline examples/structural_mechanics/crash/train.py]: You should not use the test samples for validation. We should have three splits: train, validation, and test

File: examples/structural_mechanics/crash/train.py [hunk 4]
```diff
@@ -109,6 +110,58 @@ def __init__(self, cfg: DictConfig, logger0: RankZeroLoggingWrapper):
...
+                val_cfg.num_samples = self.num_validation_samples
+            val_dataset = instantiate(
+                val_cfg,
+                name="crash_test",
+                reader=reader,
+                split="test",
```
Comment (mnabian) [inline examples/structural_mechanics/crash/train.py]: validation

File: examples/structural_mechanics/crash/train.py [hunk 5]
```diff
@@ -199,6 +252,50 @@ def backward(self, loss):
...
+            MSE_w_time += torch.mean(SqError, dim=(1, 2))
+            MSE += torch.mean(SqError)
+
+        # Sum errors across all ranks
+        if self.dist.world_size > 1:
+            torch.distributed.all_reduce(MSE, op=torch.distributed.ReduceOp.SUM)
+            torch.distributed.all_reduce(MSE_w_time, op=torch.distributed.ReduceOp.SUM)
+
+        val_stats = {
+            "MSE_w_time": MSE_w_time / self.num_validation_samples,
```
Comment (mnabian) [inline examples/structural_mechanics/crash/train.py:294]: See greptile's comment. We should divide by the actual number of validation samples, especially because you have `drop_last=True`.

Comment (dakhare-creator) [reply examples/structural_mechanics/crash/train.py:294]: Following code added at line 114-121 computes actual number of validation samples for all possible cases-
            self.num_validation_replicas = min(
                self.dist.world_size, cfg.training.num_validation_samples
            )
            self.num_validation_samples = (
                cfg.training.num_validation_samples
                // self.num_validation_replicas
                * self.num_validation_replicas
            )
            
Eg: if user provide: world_size = 3, num_validation_samples = 8, 
then num_validation_samples = 6. Each 3 node will load 2 sample.
if user provide: world_size = 24, num_validation_samples = 8, 
then num_validation_samples = 8. Only 8 node will load 1 sample and rest will have no data for validation.

File: examples/structural_mechanics/crash/train.py [hunk 6]
```diff
@@ -109,6 +110,49 @@ def __init__(self, cfg: DictConfig, logger0: RankZeroLoggingWrapper):
         )
         self.sampler = sampler
 
+        if cfg.training.num_validation_samples > 0:
+            self.num_validation_replicas = min(self.dist.world_size, cfg.training.num_validation_samples)
+            self.num_validation_samples  = cfg.training.num_validation_samples // self.num_validation_replicas * self.num_validation_replicas
```
[the remark this answers is not in the document]
Comment (dakhare-creator) [reply examples/structural_mechanics/crash/train.py]: Following code 
self.num_validation_replicas = min(self.dist.world_size, cfg.training.num_validation_samples)
enforce num_validation_replicas >= 0 as self.dist.world_size >= 1

File: examples/structural_mechanics/crash/train.py [hunk 7]
```diff
@@ -199,6 +243,48 @@ def backward(self, loss):
...
+
+            # Compute and add error
+            SqError = torch.square(pred_seq - exact_seq)
+            MSE_w_time += torch.mean(SqError, dim=(1,2))
+            MSE += torch.mean(SqError)
+
+        # Sum errors across all ranks
+        if self.dist.world_size > 1:
+            torch.distributed.all_reduce(MSE, op=torch.distributed.ReduceOp.SUM)
+            torch.distributed.all_reduce(MSE_w_time, op=torch.distributed.ReduceOp.SUM)
```
[the remark this answers is not in the document]
Comment (dakhare-creator) [reply examples/structural_mechanics/crash/train.py:291]: Following code added at line 114-121 adjust the number of validation samples for all possible cases-
self.num_validation_replicas = min(
self.dist.world_size, cfg.training.num_validation_samples
)
self.num_validation_samples = (
cfg.training.num_validation_samples
// self.num_validation_replicas
* self.num_validation_replicas
)

Eg: if user provide: world_size = 3, num_validation_samples = 8,
then num_validation_samples = 6. Each 3 node will load 2 sample.
if user provide: world_size = 24, num_validation_samples = 8,
then num_validation_samples = 8. Only 8 node will load 1 sample and rest will have no data for validation.

File: examples/structural_mechanics/crash/train.py (hunk 3 above)
Comment (dakhare-creator) [reply examples/structural_mechanics/crash/train.py]: Validation path created.

File: examples/structural_mechanics/crash/train.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mnabian) [reply examples/structural_mechanics/crash/train.py:291]: greptile, does this address your comment?

**Review by mnabian — APPROVED** [review]
LGTM! Thanks for addressing the comments!
